instructionAPI/src/InstructionDecoder-power.C

```diff
@@ -294,28 +294,159 @@ which are both 0).
       /*********************************************************************
-       *  Power ISA Version 3.1, Book I. May 2020. 2.4 Branch Instructions
+       *  Power ISA Version 3.1C, Book I. May 2024
+       *  2.4 Branch Instructions
+       *
+       *  bclr   BO,BI,BH (LK=0)
+       *  bclrl  BO,BI,BH (LK=1)
+       *
+       *  Extended form examples:
+       *    bclr  4,6
+       *    bltlr
+       *    bnelr cr2
+       *    bdnzlr
+       *
+       *
+       *  Programming Note (pg 38)
+       *  ------------------------
+       *
+       *  1. Do not use bclrl as a subroutine call
+       *
+       *  2. The conventional return idiom is to use `bclr` with BH=0b00
        *
-       *  The conventional return sequence is to use `bclr` (Branch
-       *  Conditional to Link Register) with BH=0b00.
        *********************************************************************/
-      auto target = makeRegisterExpression(ppc32::lr);
-      insn_in_progress->addSuccessor(std::move(target), false, true, bcIsConditional, false);
+      bool const LK = field<31, 31>(insn);
+
+      // link register is _always_ read, but conditionally written
+      {
+        auto link_register = makeRegisterExpression(ppc32::lr);
+        constexpr bool is_read = true;
+        constexpr bool is_implicit = true;
+        bool const is_written = (LK == 1);
+        insn_in_progress->appendOperand(std::move(link_register), is_read, is_written, is_implicit);
+      }
+
+      if(LK == 1) {
+        // 1. Do not use bclrl as a subroutine call
+        constexpr bool is_call = false;
+
+        auto lr = makeRegisterExpression(ppc32::lr);
+        constexpr bool is_indirect = true;
+        constexpr bool is_conditional = true;
+        constexpr bool is_fallthrough = false;
+        insn_in_progress->addSuccessor(std::move(lr), is_call, is_indirect, is_conditional, is_fallthrough);
+      }
+      else {
+        // 2. The conventional return sequence is to use `bclr` with BH=0b00
+        bool const is_subroutine_return = [this]() {
+          auto const bh_field = field<19,20>(insn);
+          return bh_field == 0;
+        }();
+
+        if(is_subroutine_return) {
+          // This is _not_ a fallthrough conditional
+          this->bcIsConditional = false;
+
+          // Indirectly branch through the link register
+          auto lr = makeRegisterExpression(ppc32::lr);
+          constexpr bool is_call = false;
+          constexpr bool is_indirect = true;
+          constexpr bool is_conditional = false;
+          constexpr bool is_fallthrough = false;
+          insn_in_progress->addSuccessor(lr, is_call, is_indirect, is_conditional, is_fallthrough);
+        }
+      }
 
       if(bcIsConditional) {
-        insn_in_progress->addSuccessor(makeFallThroughExpr(), false, false, false, true);
+        auto target = makeFallThroughExpr();
+        constexpr bool is_call = false;
+        constexpr bool is_indirect = true;
```

Comment:
I assume fallthrough should be indirect as in the original code?

Reply:
It's unclear what "indirect" means here. Except for printing, Dyninst only uses that information in only one place: https://github.com/dyninst/dyninst/blob/v13.0.0/dyninstAPI/src/Relocation/Widgets/CFWidget.C#L150. 

The comment there is

```
   // Indirect branch:
   //   1) Just go for it... we have no control, really
```

Super helpful. Given this, I interpret an "indirect" control flow target to be contained in some location that must be dereferenced before being used (e.g., branching through a register) as opposed to being a literal offset or absolute target. Hence, fallthrough targets are always indirect since they go through the PC (or the LR for ppc).

Reply:
I thiink fallthrough targets are always direct snice they are literally the next instruction and does not need to go through the process of resolving indirect jump target.

Reply:
My biggest problem is that `InstructionDecoder_power::makeFallThroughExpr` assumes that the fallthrough target can be addressed via the PC, but Power doesn't have one[1]. Since it's not direct, then I considered it indirect. But a re-read of the decode algorithms for conditional branch instructions doesn't actually indicate that. I'll change this back to being direct to make it consistent with the other architectures. As an aside, I plan to change the  CFT interface when updating to Capstone to make all of this more explicit.

[1] From the docs:
```
Obtaining the address of the next instruction:
Use the following form of Branch and Link.

    bcl 20,31,$+4
```
where `$` is an assembler directive.